File: 39. Combination Sum/39. Combination Sum.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        ans, curr = [], []

        def backtrack(start, curr_sum):
            if curr_sum == target:
                ans.append(curr[:])
                return

            if curr_sum > target:
                return

            for i in range(start, n):
                curr.append(candidates[i])
                backtrack(i, curr_sum + candidates[i])
                curr.pop()

        backtrack(0, 0)
        return ans

File: 39. Combination Sum/39. Combination Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  n = candidates.length;
  const ans = [];
  const curr = [];

  const backtrack = (start, currSum) => {
    if (currSum === target) {
      ans.push([...curr]);
      return;
    }

    if (currSum > target) {
      return;
    }

    for (let i = start; i < n; i++) {
      curr.push(candidates[i]);
      backtrack(i, currSum + candidates[i]);
      curr.pop();
    }
  };

  backtrack(0, 0);
  return ans;
};
